File: sdk/python/tests/e2e_control.py
"""E2E: control surface — delay/promote, moves, updates, clean, retry."""

from __future__ import annotations

import time

from harness import Server, test, unique_name, wait_until

from bunqueue import Queue, Worker


@test
def delayed_then_promote(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {}, delay=60000)
        assert queue.get_state(job.id) == "delayed"
        queue.promote_job(job.id)
        assert queue.get_state(job.id) in ("waiting", "prioritized")


@test
def promote_jobs_bulk(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        for i in range(3):
            queue.add("t", {"i": i}, delay=60000)
        # PromoteJobs scans persisted delayed jobs: wait for write-buffer flush
        assert wait_until(lambda: len(queue.get_delayed()) == 3, 10)
        promoted = queue.promote_jobs()
        assert promoted == 3, f"promoted {promoted}"
        assert queue.get_waiting_count() == 3


@test
def move_to_delayed_and_change_delay(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {})
        queue.move_job_to_delayed(job.id, 60000)
        assert queue.get_state(job.id) == "delayed"
        queue.change_job_delay(job.id, 1)
        assert wait_until(lambda: queue.get_state(job.id) in ("waiting", "prioritized"), 10)


@test
def update_data_and_priority(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {"v": 1})
        queue.update_job_data(job.id, {"name": "t", "v": 2})
        fetched = queue.get_job(job.id)
        assert fetched is not None and fetched.data["v"] == 2
        queue.change_job_priority(job.id, 99)
        fetched = queue.get_job(job.id)
        assert fetched is not None and fetched.priority == 99


@test
def remove_and_discard(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("t", {})
        queue.remove(job.id)
        assert queue.get_job(job.id) is None or queue.get_state(job.id) not in (
            "waiting",
            "prioritized",
        )
        assert queue.count() == 0


@test
def drain_removes_waiting(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        queue.add_bulk([{"name": "t", "data": {"i": i}} for i in range(10)])
        assert queue.count() == 10
        queue.drain()
        assert queue.count() == 0


@test
def obliterate_wipes_queue(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        queue.add_bulk([{"name": "t", "data": {}} for _ in range(5)])
        queue.pause()
        queue.obliterate()
        assert queue.count() == 0
        assert queue.is_paused() is False  # obliterate resets state


@test
def retry_failed_job(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        job = queue.add("boom", {}, attempts=1)
        worker = Worker(
            queue.name,
            lambda j: (_ for _ in ()).throw(RuntimeError("nope")),
            port=server.port,
            poll_timeout_ms=500,
        )
        worker.start()
        try:
            assert wait_until(lambda: queue.get_state(job.id) == "failed", 15)
        finally:
            worker.close(timeout=10)
        # failed -> waiting (BullMQ retry contract)
        queue.retry_job(job.id)
        assert queue.get_state(job.id) in ("waiting", "prioritized")


@test
def clean_completed(server: Server) -> None:
    with Queue(unique_name("q"), port=server.port) as queue:
        done = []
        worker = Worker(queue.name, lambda j: "ok", port=server.port, poll_timeout_ms=500)
        worker.on("completed", lambda j, r: done.append(j.id))
        worker.start()
        try:
            for i in range(4):
                queue.add("t", {"i": i})
            assert wait_until(lambda: len(done) == 4, 15)
        finally:
            worker.close(timeout=10)
        time.sleep(0.2)
        removed = queue.clean(0, state="completed")
        assert removed == 4, f"cleaned {removed}"


@test
def lifo_ordering(server: Server) -> None:
    """LIFO applies among lifo jobs: newest first (mixed falls back to FIFO)."""
    with Queue(unique_name("q"), port=server.port) as queue:
        queue.add("first", {"o": 1}, lifo=True)
        queue.add("second", {"o": 2}, lifo=True)
        pulled = queue.connection.call(
            {"cmd": "PULL", "queue": queue.name, "owner": "e2e", "timeout": 1000}
        )
        assert pulled["job"]["data"]["o"] == 2, f"lifo not respected: {pulled['job']['data']}"

File: sdk/typescript/tests/e2e-control.ts
/** E2E: control surface — delay/promote, moves, updates, clean, retry, lifo. */

import type { Job } from '../dist/legacy.js';
import { assert, assertEq, makeQueue, makeWorker, sleep, test, waitFor } from './harness.ts';

test('control: delayed + promote single', async () => {
  const queue = makeQueue('delay');
  try {
    const job = await queue.add('later', { x: 1 }, { delay: 60_000 });
    assertEq(await queue.getJobState(job.id), 'delayed', 'job is delayed');
    await queue.promoteJob(job.id);
    const state = await queue.getJobState(job.id);
    assert(state === 'waiting' || state === 'prioritized', `promoted (${state})`);
  } finally {
    queue.close();
  }
});

test('control: promoteJobs bulk returns count', async () => {
  const queue = makeQueue('promoteb');
  try {
    for (let i = 0; i < 3; i++) await queue.add('t', { i }, { delay: 60_000 });
    // PromoteJobs scans persisted delayed jobs: wait for write-buffer flush
    await waitFor(async () => (await queue.getDelayed()).length === 3, 10_000);
    const promoted = await queue.promoteJobs();
    assertEq(promoted, 3, 'promoted count');
    assertEq(await queue.getWaitingCount(), 3, 'all waiting');
  } finally {
    queue.close();
  }
});

test('control: moveToDelayed + changeDelay', async () => {
  const queue = makeQueue('movedelay');
  try {
    const job = await queue.add('t', {});
    await queue.moveJobToDelayed(job.id, 60_000);
    assertEq(await queue.getJobState(job.id), 'delayed', 'moved to delayed');
    await queue.changeJobDelay(job.id, 1);
    await waitFor(async () => {
      const state = await queue.getJobState(job.id);
      return state === 'waiting' || state === 'prioritized';
    }, 10_000);
  } finally {
    queue.close();
  }
});

test('control: updateJobData + changeJobPriority', async () => {
  const queue = makeQueue('mutate');
  try {
    const job = await queue.add('t', { v: 1 });
    await queue.updateJobData(job.id, { name: 't', v: 2 });
    const fetched = await queue.getJob(job.id);
    assertEq((fetched?.data as Record<string, unknown> | undefined)?.v, 2, 'data updated');
    await queue.changeJobPriority(job.id, { priority: 99 });
    const again = await queue.getJob(job.id);
    assertEq(again?.priority, 99, 'priority changed');
  } finally {
    queue.close();
  }
});

test('control: remove job', async () => {
  const queue = makeQueue('remove');
  try {
    const job = await queue.add('t', {});
    await queue.remove(job.id);
    assertEq(await queue.getJob(job.id), null, 'job gone');
    assertEq(await queue.count(), 0, 'count 0');
  } finally {
    queue.close();
  }
});

test('control: drain removes waiting', async () => {
  const queue = makeQueue('drain');
  try {
    await queue.addBulk(Array.from({ length: 10 }, (_, i) => ({ name: 't', data: { i } })));
    assertEq(await queue.count(), 10, 'ten queued');
    await queue.drain();
    assertEq(await queue.count(), 0, 'drained');
  } finally {
    queue.close();
  }
});

test('control: obliterate wipes queue and resets pause', async () => {
  const queue = makeQueue('oblit');
  try {
    await queue.addBulk(Array.from({ length: 5 }, () => ({ name: 't', data: {} })));
    await queue.pause();
    await queue.obliterate();
    assertEq(await queue.count(), 0, 'wiped');
    assertEq(await queue.isPaused(), false, 'obliterate resets paused state');
  } finally {
    queue.close();
  }
});

test('control: pause / resume / isPaused', async () => {
  const queue = makeQueue('pause');
  try {
    await queue.pause();
    assertEq(await queue.isPaused(), true, 'paused');
    await queue.resume();
    assertEq(await queue.isPaused(), false, 'resumed');
  } finally {
    queue.close();
  }
});

test('control: retryJob failed → waiting', async () => {
  const queue = makeQueue('retryjob');
  const worker = makeWorker(queue.name, async () => {
    throw new Error('nope');
  });
  try {
    const job = await queue.add('boom', {}, { attempts: 1 });
    await waitFor(async () => (await queue.getJobState(job.id)) === 'failed');
    await worker.close();
    await queue.retryJob(job.id);
    const state = await queue.getJobState(job.id);
    assert(state === 'waiting' || state === 'prioritized', `retried to waiting (${state})`);
  } finally {
    if (!worker.isClosed()) await worker.close();
    queue.close();
  }
});

test('control: clean completed returns count', async () => {
  const queue = makeQueue('clean');
  const done: string[] = [];
  const worker = makeWorker(queue.name, async () => 'ok');
  worker.on('completed', (job: Job) => done.push(job.id));
  try {
    for (let i = 0; i < 4; i++) await queue.add('t', { i });
    await waitFor(() => done.length === 4);
    await worker.close();
    await sleep(200);
    const removed = await queue.clean(0, undefined, 'completed');
    assertEq(removed, 4, 'cleaned count');
  } finally {
    if (!worker.isClosed()) await worker.close();
    queue.close();
  }
});

test('control: lifo ordering among lifo jobs', async () => {
  const queue = makeQueue('lifo');
  try {
    // LIFO applies among lifo jobs (mixed with FIFO jobs falls back to FIFO)
    await queue.add('first', { o: 1 }, { lifo: true });
    await queue.add('second', { o: 2 }, { lifo: true });
    const response = await queue.connection.call({
      cmd: 'PULL',
      queue: queue.name,
      owner: 'e2e',
      timeout: 1000,
    });
    const pulled = response.job as Record<string, unknown>;
    assertEq((pulled.data as Record<string, unknown>).o, 2, 'newest lifo job first');
  } finally {
    queue.close();
  }
});

test('control: priority ordering via PULL', async () => {
  const queue = makeQueue('prio');
  try {
    await queue.add('low', { p: 'low' }, { priority: 1 });
    await queue.add('high', { p: 'high' }, { priority: 100 });
    const response = await queue.connection.call({
      cmd: 'PULL',
      queue: queue.name,
      owner: 'e2e',
      timeout: 1000,
    });
    const pulled = response.job as Record<string, unknown>;
    assertEq((pulled.data as Record<string, unknown>).p, 'high', 'high priority first');
  } finally {
    queue.close();
  }
});
